File: java/src/main/java/com/example/cote/PGMRS/단어_변환.java
package com.example.cote.PGMRS;


import java.util.*;
import java.io.*;
import java.nio.file.*;

public class 단어_변환 {	

	private static final int MAX_WORDS_LENGTH = 10;

	public static record State (String word, int step){}

	private static boolean isConvertable(String src, String dst) {
		
		char[] srcArr = src.toCharArray();
		char[] dstArr = dst.toCharArray();

		int diff = 0;

		for (int i = 0; i < srcArr.length; i++) {
			if (srcArr[i] != dstArr[i]) diff++;
		}

		return diff == 1;
	}

	private static int solve(String begin, String target, String[] words) {

		boolean[] isVisited = new boolean[words.length];

		Queue<State> queue =new LinkedList<> ();
		queue.add(new State(begin, 0));

		while (!queue.isEmpty()) {
			State state = queue.poll();

			if (state.word.equals(target)){
				return state.step;
			}

			for (int i = 0; i < words.length; i++) {
				String next = words[i];

				if (!isConvertable(state.word, next)){
					continue;
				}

				if (isVisited[i]) {
					continue;
				}

				isVisited[i] = true;
				queue.add(new State(next, state.step + 1));
			}
		}
		return 0;
	}

	public static void main(String... args) throws IOException{

		Path p = Paths.get(System.getProperty("user.dir")+"/data/단어_변환.txt");
		BufferedReader reader = Files.newBufferedReader(p);

		int testCases = Integer.parseInt(reader.readLine());

		while (testCases > 0) {
			String begin = reader.readLine();
			String target = reader.readLine();

			String[] words = reader.readLine().split(" ");
			int answer = Integer.parseInt(reader.readLine());

			System.out.printf("%d %d\n", answer, solve(begin, target, words));

			testCases--;
		}
	}
}

File: kotlin/src/main/kotlin/com/example/cote/PGMRS/단어_변환.kt
package com.example.cote.PGMRS

import java.io.*
import java.nio.file.*

data class State( 
	val word: String,
	val step : Int
	){}

fun isConvertable(word: String, next: String): Boolean =
	IntRange(0, word.length - 1)
		.count{i -> word[i] != next[i]} == 1

fun solve(begin: String, target: String, words: List<String>): Int {
	val visited = MutableList(words.size) {false}
	val queue = ArrayDeque<State>()
	queue.add(State(begin, 0))

	while (!queue.isEmpty()) {
		val (word, step) = queue.removeFirst()

		if (word.equals(target)) {
			return step
		}

		for ((i, next) in words.withIndex()) {
			if (visited[i]) continue;
			if (isConvertable(word, next)) {
				visited[i] = true
				queue.add(State(next, step + 1))
			}
		}
	}

	return 0
}

fun main() {
	val p = Paths.get(System.getProperty("user.dir" )+ "/data/단어_변환.txt");
	val reader = Files.newBufferedReader(p);

	repeat (reader.readLine().toInt()) {
		val begin = reader.readLine()
		val target = reader.readLine()
		val words = reader.readLine().split(" ")
		val answer = reader.readLine().toInt()

		println("$answer ${solve(begin, target, words)}")
	}
}